File: algorithms/c/234.c
#include <stdio.h>
#include <stdbool.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *reverse(struct ListNode *head)
{
    if ( head == NULL )
        return head;
    struct ListNode *newHead = NULL, *tmp = NULL;
    while ( head )
    {
        tmp = head;
        head = head->next;
        tmp->next = newHead;
        newHead = tmp;
    }
    return newHead;
}

bool isPalindrome(struct ListNode* head)
{
    if ( head == NULL )
        return true;
    struct ListNode *slow = head, *fast = head;
    while ( fast->next != NULL && fast->next->next != NULL )
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    slow->next = reverse(slow->next);
    slow = slow->next;
    while ( slow != NULL )
    {
        if ( head->val != slow->val )
            return false;
        head = head->next;
        slow = slow->next;
    }
    return true;
}

int main()
{
    return 0;
}

File: algorithms/python/234.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        def findMid(head):
            slow, fast = head, head
            while slow and fast and fast.next:
                slow = slow.next
                fast = fast.next.next
            return slow
        def rev(head):
            if not head: return None
            rls, head = head, head.next
            rls.next = None
            while head:
                node, head = head, head.next
                rls, node.next = node, rls
            return rls

        if not head: return True
        mid = findMid(head)
        mid = rev(mid)
        while mid:
            if head.val != mid.val:
                return False
            head, mid = head.next, mid.next
        return True
